### PR TITLE
Fuzz Testing: Clarity value sanitization and serialization

### PR DESCRIPTION
**Pull request by kantai**
### Description

This PR adds a `cargo fuzz` target that instruments and makes some assertions about the 2.4-enabled clarity value sanitization as well as existing serialization and deserialization routines.

Unlike previous fuzzing PRs, this one does not need to make any changes in the `clarity` module: implementation of `Arbitrary` is done in the fuzzing module itself.

### Additional info (benefits, drawbacks, caveats)

This PR (apart from a `warn -> debug` log downgrade) does not make any changes to the codebase, it just adds new testing coverage. This could also be used a starting point for future fuzz testing (see discussion #3732)

### CODE REVIEW

File: clarity/fuzz/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+[profile.release]
+debug = 1
+
+[[bin]]
+name = "fuzz_sanitize"
+path = "fuzz_targets/fuzz_sanitize.rs"
+test = false
+doc = false
+
+# [[bin]]
```
Comment (jcnelson) [inline clarity/fuzz/Cargo.toml:35]: nit: Are these just TODOs, or are the artifacts of a change that is no longer meant to be included? Either way, commented-out code is confusing.

File: clarity/fuzz/fuzz_targets/fuzz_sanitize.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
+#![no_main]
```
Comment (jcnelson) [inline clarity/fuzz/fuzz_targets/fuzz_sanitize.rs:1]: copyright statement is missing

File: clarity/fuzz/fuzz_targets/fuzz_sanitize.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+                    .into()
+            },
+            _ => return Err(arbitrary::Error::IncorrectFormat),
+        };
+
+        Ok(FuzzClarityValue(clar_value))
+    }
+}
+
+pub fn strict_admits(me: &TypeSignature, x: &ClarityValue) -> Result<bool, CheckErrors> {
```
Comment (jcnelson) [inline clarity/fuzz/fuzz_targets/fuzz_sanitize.rs:117]: I notice there's no recursion depth checking here.  Could this lead to a stack overflow, given that both `me` and `x` are generated from the fuzzer?

File: clarity/fuzz/fuzz_targets/fuzz_sanitize.rs [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+fuzz_target!(|value: FuzzClarityValue| {
+    fuzz_sanitize(value.0);
+});
+
+
+
+/// Same as fuzz_sanitize, but does not check any serialization routines
+fn fuzz_value_sanitize(input: ClarityValue) {
+    let computed_type = TypeSignature::type_of(&input);
```
Comment (jcnelson) [inline clarity/fuzz/fuzz_targets/fuzz_sanitize.rs:236]: Can `TypeSignaure::type_of` OOM if `input` is too deep?

File: clarity/fuzz/fuzz_targets/fuzz_sanitize.rs (hunk 3 above)
Comment (kantai) [reply clarity/fuzz/fuzz_targets/fuzz_sanitize.rs:117]: Like below, it shouldn't. `TypeSignature` construction (which uses the `TypeSignature::type_of()` method) will not construct a deep enough type signature. The tuple bug fixed in 2.4 also cannot trigger a stack overflow here -- the length is compared before descent.

If it did result in a stack overflow, the fuzzer would (correctly) raise the input case as a failure case.

File: clarity/fuzz/fuzz_targets/fuzz_sanitize.rs (hunk 2 above)
Comment (kantai) [reply clarity/fuzz/fuzz_targets/fuzz_sanitize.rs:16]: Added

File: clarity/fuzz/Cargo.toml (hunk 1 above)
Comment (kantai) [reply clarity/fuzz/Cargo.toml]: Removed -- these were artifacts from the prior fuzzing draft PRs.

**Review by jcnelson — APPROVED** [review]
Thanks for addressing my feedback!

**Review by obycode — APPROVED** [review]
LGTM!

**Pull request merged**